src/main/java/com/dev/cinema/dao/impl/MovieSessionDaoImpl.java

```diff
@@ -0,0 +1,69 @@
+package com.dev.cinema.dao.impl;
+
+import com.dev.cinema.dao.MovieSessionDao;
+import com.dev.cinema.exception.DataProcessingException;
+import com.dev.cinema.lib.Dao;
+import com.dev.cinema.model.MovieSession;
+import com.dev.cinema.util.HibernateUtil;
+import java.time.LocalDate;
+import java.time.LocalTime;
+import java.util.List;
+import javax.persistence.criteria.CriteriaQuery;
+import org.hibernate.Session;
+import org.hibernate.Transaction;
+
+@Dao
+public class MovieSessionDaoImpl implements MovieSessionDao {
+    @Override
+    public MovieSession add(MovieSession movieSession) {
+        Transaction transaction = null;
+        Session session = null;
+        try {
+            session = HibernateUtil.getSessionFactory().openSession();
+            transaction = session.beginTransaction();
+            session.persist(movieSession);
+            transaction.commit();
+            return movieSession;
+        } catch (Exception e) {
+            if (transaction != null) {
+                transaction.rollback();
+            }
+            throw new DataProcessingException("Can not insert movie session: "
+                    + movieSession.getMovie(), e);
+        } finally {
+            if (session != null) {
+                session.close();
+            }
+        }
+    }
+
+    @Override
+    public List<MovieSession> getAll() {
+        try (Session session = HibernateUtil.getSessionFactory().openSession()) {
+            CriteriaQuery<MovieSession> criteriaQuery = session.getCriteriaBuilder()
+                    .createQuery(MovieSession.class);
+            criteriaQuery.from(MovieSession.class);
+            return session.createQuery(criteriaQuery).getResultList();
+        } catch (Exception e) {
+            throw new DataProcessingException("Error occurred while retrieving movie sessions", e);
+        }
+    }
+
+    @Override
+    public List<MovieSession> findAvailableSessions(Long movieId, LocalDate date) {
+        try (Session session = HibernateUtil.getSessionFactory().openSession()) {
+            return session
+                    .createQuery("SELECT ms FROM MovieSession ms "
+                            + "WHERE ms.id = :id "
+                            + "AND ms.showTime >= :timeFrom "
+                            + "AND ms.showTime <= :timeTo", MovieSession.class)
```

Comment:
I made this with fetch joins. Don't know if it is necessary here.

Reply:
As I understood from QnA - both options are valid, if model relations is established correctly everything works just fine